content/ui-patterns/saving.mdx

```diff
@@ -0,0 +1,222 @@
+    <Caption>Mix auto-saving controls with explictly saved controls in the same form</Caption>
+  </Dont>
+</DoDontContainer>
+
+### Unsaved changes
+
+With the addition of the command palette there are additional opportunities to accidentally navigate away from a page with unpublished edits.
+
+If a user tries to navigate away from a page with a form that has unsaved changes, you can optionally warn them that their changes will be lost using the [`beforeunload` event](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event). Modern browsers prevent developers from customizing the `alert` message, so we're limited to a generic message like "Do you want to leave this site? Changes you made may not be saved."
+
+### Declarative controls
+
+- Text inputs
+- Checkbox groups
+- Radio button groups
+- Browser-native `<select>` dropdown menus
+- Multi-select dropdown menus
+
+To maintain consistent interaction patterns across your app, do not use segmented controls in an explicitly saved form. These should only be used with automatic saving.
+
+To ensure your forms are accessible, do not automatically save declarative inputs.
+
+#### Problems with autosaving declarative inputs:
+**Text inputs**
+- Users are expecting to set a value and then submit
+- Sensitive information could be unintentionally submitted. For example: a change password input
+
+**Checkbox groups**
+- It could be unclear whether or not their selections have been saved
+- Users may accidentally submit a selection that they clicked by accident
+
+**Radio button groups**
+- Screenreader users can't read the options without selecting them. This could accidentally apply a selection the user didn't want.
+
+**Browser-native `<select>` dropdowns**
+- A select dropdown has a similar problem as a radio button group. Options are selected when a user focuses the select and uses their arrow keys to read through each option
+
+<DoDontContainer>
+  <Do>
+    <img src="https://user-images.githubusercontent.com/2313998/151623711-ad294946-fd67-428d-96b2-e4f534857c8a.png" />
+    <Caption>Use save and cancel buttons in dropdowns that take multiple selections</Caption>
+  </Do>
+  <Dont>
+    <img src="https://user-images.githubusercontent.com/2313998/151623713-12028e6d-86ba-4ae6-b46b-d3c88c0fa2c0.png" />
+    <Caption>Automatically save selections when the dropdown is closed</Caption>
+  </Dont>
+</DoDontContainer>
+
+## Automatic saving
+
+Automatic saving applies changes as they are made.
+
+Automatic saving should be used when the user expects instant feedback from the change they made. Changes should only be saved automatically using [imperative inputs](#imperative-inputs).
+
+<img
+    width="960"
+    src="https://user-images.githubusercontent.com/2313998/151887933-bc96de6a-bc80-43d5-a66a-54a7b202eae3.png"
+    alt="Two images of the single-select dropdown for a 'milestone' on a GitHub issue"
+/>
```

Comment:
I'd suggest a different visual example on this one. Even though it's something that's just one selection, it also contains a filter and tabs, which means it's still technically a dialog and would need the save / cancel 😬. Single-select "menus" can't have these extra things. James mentions it in this issue: https://github.com/github/accessibility/issues/507#issuecomment-1018612516

Can we use something like the Unwatch menu? (doesn't have to be that but a true menu that doesn't have anything extra like a filter or tabs.

Reply:
Thanks for sharing that comment @ichelsea!

I want to update these guidelines to say that even single-select dropdowns require "save" and "cancel" buttons if they have a filter or tabs, but I don't understand the reason for that. Could you help me understand?

Reply:
Yep! So, simple single select menus cannot contain any added features like search inputs / filters or tabs. The minute those things are added, they no longer are a semantic menu and need to be placed in a dialog for proper navigation and semantics. A semantic menu can only be a list of items that can be selected.

I would say that it's definitely not _as big_ of an issue as it is with multi-select but it's still an issue.

Reply:
Thanks for the explanation! I'll add that info.

Reply:
@ichelsea - I made my first attempt at adding guidance for only auto-saving semantic menus: https://github.com/primer/design/pull/212/files#diff-178947cd6d73ea6bca9fb984bf99e45488dfd46d44df47f319b1a214235dc026R230-R234

Let me know if this sounds correct to you.

PS: I did this at the end of the day, so it's probably a little sloppy and needs rewording.